### PR TITLE
Inplace date range picker

### PR DESCRIPTION
**Pull request by NobodysNightmare**
## OpenProject work package

http://community.openproject.org/work_packages/18939
## Notes

This PR is a re-issue of https://github.com/opf/openproject/pull/2662 that has the accidential commit of the `Gemfile.lock` removed.

Due to time constraints the following issues are not handled in this PR, but get separate work packages:
- [date range picker is closed on scrolling](https://community.openproject.org/work_packages/19612)
- [clicking `Clear` does not close the date range picker](https://community.openproject.org/work_packages/19613)
- [it is not possible to edit date custom fields via inplace edit](https://community.openproject.org/work_packages/19614)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-inplace-date-range-picker.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-inplace-date-range-picker.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by NobodysNightmare** [comment]
Protractor tests failed two times in a row. If they don't turn green now (third try), I'll consider them as correctly failing...

**Comment by floriank** [comment]
> ERROR - Unable to start a WebDriver session.

Does not seem like a test issue per se. @myabc any ideas ?

File: frontend/tests/integration/protractor.conf.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ exports.config = {
 
   // Capabilities to be passed to the webdriver instance.
   capabilities: {
-    'browserName': 'firefox'
+    'browserName': 'chrome'
```
Comment (myabc) [inline frontend/tests/integration/protractor.conf.js]: @hirtie-maxim this shouldn't be changed.

**Comment by myabc** [comment]
@floriank 

> Does not seem like a test issue per se. @myabc any ideas ?

See https://github.com/opf/openproject/pull/2817/files#r28232028

Travis doesn't support Chrome.

**Comment by floriank** [comment]
@NobodysNightmare would you be so kind and revert the `protractor.conf` ?

nvm. I noticed you just added me as collab.

File: frontend/app/config/configuration-service.js [hunk 2]
```diff
@@ -90,5 +90,11 @@ module.exports = function() {
     isModuleEnabled: function(module) {
       return this.settings.enabled_modules.indexOf(module) >= 0;
     },
+    startOfWeekPresent: function() {
+      return this.displaySettingPresent('start_of_week') && this.setting.display.start_of_week != '';
```
Comment (floriank) [inline frontend/app/config/configuration-service.js]: this was introduced by @hschink intiall, nothing @NobodysNightmare , @hirtie-maxim or myself touched.

Comment (NobodysNightmare) [inline frontend/app/config/configuration-service.js]: Lets wait if he returns to fix that styling issue xD

File: .jshintignore [hunk 3]
```diff
@@ -4,3 +4,5 @@ app/assets/javascripts/jstoolbar/**/*
 app/assets/javascripts/jquery_noconflict.js
 app/assets/javascripts/project/**/*
 app/assets/javascripts/lib/**/*
+frontend/tests/*
+frontend/tests/**/*
```
Comment (NobodysNightmare) [inline .jshintignore]: Is this something we really want?

Isn't JSHint something like rubocop?

**Comment by NobodysNightmare** [comment]
@myabc @floriank Do you see any other file changes that should probably not be part of this PR?

I don't know all the configuration files one might accidentially change as frontend developer ^^ (similar to `Gemfile.lock`, `production.rb`, etc.)

File: .jshintignore (hunk 3 above)
Comment (myabc) [inline .jshintignore]: @NobodysNightmare right. We shouldn't be ignoring tests.

File: frontend/app/services/timezone-service.js [hunk 4]
```diff
@@ -1,30 +1,3 @@
-//-- copyright
```
Comment (myabc) [inline frontend/app/services/timezone-service.js:1]: @hirtie-maxim why did you delete the copyright notice?

File: .jshintignore (hunk 3 above)
Comment (floriank) [inline .jshintignore]: It is - this should probably also be reverted. However, I do remember, that jshint complaints are a thing with the tests, so therefore we could leave it be.

File: frontend/app/config/configuration-service.js [hunk 5]
```diff
@@ -90,5 +90,11 @@ module.exports = function() {
...
+    },
+    startOfWeek: function() {
+      return this.settings.display.start_of_week;
+    }
   };
-};
+};
```
Comment (floriank) [inline frontend/app/config/configuration-service.js]: Files should end in a newline

File: frontend/app/services/timezone-service.js [hunk 6]
```diff
@@ -52,8 +25,18 @@ module.exports = function(ConfigurationService, I18n) {
...
+    isValid: function(date) {
+      var format = ConfigurationService.dateFormatPresent() ?
+                   ConfigurationService.dateFormat() : 'L';
+      return moment(date, [format]).isValid();
     }
   };
 
   return TimezoneService;
-};
+};
```
Comment (floriank) [inline frontend/app/services/timezone-service.js]: hhhnnhnhnhgrl.

File: frontend/app/ui_components/date/date-range-picker-directive.js [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+      });
+
+      angular.element(window).resize(function() {
+        inputStart.datepicker('hide');
+        inputEnd.datepicker('hide');
+        angular.element('#ui-datepicker-div').blur();
+      });
+    }
+  };
+};
```
Comment (floriank) [inline frontend/app/ui_components/date/date-range-picker-directive.js]: same here.

File: frontend/public/templates/components/inplace_editor/date/date_range_picker.html [hunk 8]
```diff
@@ -0,0 +1,3 @@
+<div class="date-range-picker">
+	<input ng-model="editableStartDate" type="text" class="start" /><span>-</span><input ng-model="editableEndDate" type="text" class="end" />
+</div>
```
Comment (floriank) [inline frontend/public/templates/components/inplace_editor/date/date_range_picker.html]: http://stackoverflow.com/questions/729692/why-should-files-end-with-a-newline

File: .jshintignore (hunk 3 above)
Comment (myabc) [inline .jshintignore]: @floriank then we should adjust `.jshintrc` or add a special comment inline (e.g. `/*jshint expr: true*/`)

File: app/assets/javascripts/application.js.erb [hunk 9]
```diff
@@ -77,6 +77,34 @@ jQuery(document).ready(function ($) {
     return false;
   });
 
+  $.extend($.datepicker, {
```
Comment (myabc) [inline app/assets/javascripts/application.js.erb]: ARGHH!! @hirtie-maxim please don't add anything new to `application.js.erb`. Adding new scripts here should be considered deprecated. `frontend/app/misc/datepicker-defaults.js` is probably the right file to append to.

File: frontend/app/templates/components/inplace_editor/date/date_range_picker.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<div class="date-range-picker">
+	<input type="text" class="start" /><span>-</span><input type="text" class="end" />
```
Comment (myabc) [inline frontend/app/templates/components/inplace_editor/date/date_range_picker.html]: @hirtie-maxim spaces not tabs please!

File: frontend/public/templates/components/inplace_editor/date/date_range_picker.html [hunk 11]
```diff
@@ -0,0 +1,3 @@
+<div class="date-range-picker">
+	<input ng-model="editableStartDate" type="text" class="start" /><span>-</span><input ng-model="editableEndDate" type="text" class="end" />
```
Comment (myabc) [inline frontend/public/templates/components/inplace_editor/date/date_range_picker.html]: @hirtie-maxim spaces not tabs please!

File: .jshintignore (hunk 3 above)
Comment (floriank) [inline .jshintignore]: I'll revert it, let's see what happens.

**Comment by floriank** [comment]
This will probably take a little bit.

File: frontend/app/work_packages/services/work-package-field-service.js [hunk 12]
```diff
@@ -254,6 +263,8 @@ module.exports = function(
         break;
       case 'User':
         displayStrategy = 'user';
```
Comment (floriank) [inline frontend/app/work_packages/services/work-package-field-service.js:265]: @hirtie-maxim there is a `break` missing here, leading to a problem where all user fields will be treated as `daterange` inputs.

File: frontend/app/misc/datepicker-defaults.js [hunk 13]
```diff
@@ -68,4 +68,33 @@ jQuery(function($) {
...
+    // Reference the orignal function so we can override it and call it later
+    _inlineDatepicker2: $.datepicker._inlineDatepicker,
+
+    // Override the _inlineDatepicker method
+    _inlineDatepicker: function (target, inst) {
+
+      // Call the original
+      this._inlineDatepicker2(target, inst);
+
+      var beforeShow = $.datepicker._get(inst, 'beforeShow');
```
Comment (floriank) [inline frontend/app/misc/datepicker-defaults.js:84]: I have no friggin idea here.

**Comment by floriank** [comment]
I cleaned everything up, no need for ignoring the frontend specs. The implementation for the bugs has to wait though, I need my beauty sleep.

**Comment by NobodysNightmare** [comment]
@myabc Can you have a look at this?
I don't feel to be able to review the JS changes @floriank made...

**Comment by floriank** [comment]
The JS changes is mostly styling, cleaning up lines and the case of the forgotten `break`  by @hirtie-maxim
